src/testrender/shading.h

```diff
@@ -32,71 +245,110 @@ struct BSDF {
         float pdf;
         float roughness;
     };
-    BSDF() {}
+    OSL_HOSTDEVICE BSDF(ClosureIDs id=EMPTY_ID) : id(id) {}
+
+#ifndef __CUDACC__
     virtual Color3 get_albedo(const Vec3& /*wo*/) const { return Color3(1); }
     virtual Sample eval(const Vec3& wo, const Vec3& wi) const = 0;
-    virtual Sample sample(const Vec3& wo, float rx, float ry, float rz) const
-        = 0;
+    virtual Sample sample(const Vec3& wo, float rx, float ry, float rz) const = 0;
+#else
+    OSL_HOSTDEVICE Color3 get_albedo(const Vec3& /*wo*/) const
+    {
+        return Color3(1);
+    }
+    OSL_HOSTDEVICE Sample eval(const Vec3& wo, const Vec3& wi) const
+    {
+        return {};
+    }
+    OSL_HOSTDEVICE Sample sample(const Vec3& wo, float rx, float ry,
+                                 float rz) const
+    {
+        return {};
```

Comment:
All this should be unreachable code right? It would be good to crash there to make sure you don't call it by accident.

Reply:
If we refactor both to be "switch dispatch" then the virtual keyword goes away and the base implementation is not needed - just a default case in the switch (which can return 0 safely).

Reply:
I'll revisit this one the switch stuff is in place.